lib/src/rules/unnecessary_nullable_for_final_variable_declarations.dart

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Use non-nullable type '
+    'for final variable initialized with non-nullable value.';
+
+const _details = r'''
+
+Use non-nullable type for final variable initialized with non-nullable value.
+
+**BAD:**
+```
+final int? i = 1;
+```
+
+**GOOD:**
+```
+final int i = 1;
+```
+
+''';
+
+class UnnecessaryNullableForFinalVariableDeclarations extends LintRule
+    implements NodeLintRule {
+  UnnecessaryNullableForFinalVariableDeclarations()
+      : super(
+            name: 'unnecessary_nullable_for_final_variable_declarations',
+            description: _desc,
+            details: _details,
+            maturity: Maturity.experimental,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+    registry.addFieldDeclaration(this, visitor);
+    registry.addTopLevelVariableDeclaration(this, visitor);
+    registry.addVariableDeclarationStatement(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  bool _isNonNullableEnabled;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    _isNonNullableEnabled = node.featureSet.isEnabled(Feature.non_nullable);
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    if (!_isNonNullableEnabled) return;
+
+    for (var variable in node.fields.variables) {
+      // We could also include public members in private classes but to do that
+      // we'd need to ensure that there are no instances of either the
+      // enclosing class or any subclass of the enclosing class that are ever
+      // accessible outside this library.
```

Comment:
I don't think that's actually the issue here. The only reason I can think of for not linting public fields is because either a superclass or subclass might require the nullable type because of inheritance semantics. We could check the signature against superclasses, but we can't know find all subclasses unless the class is private (because then all of the subclasses will be in the same library).

We should update the comment, but I'm _not_ recommending that we do the extra work to lint public members in private classes. I don't think that case will come up often enough to be worth the effort, and we can always add it later if I'm wrong.

Reply:
Comment removed.